pkg/chartsync/git.go

```diff
@@ -0,0 +1,407 @@
+	if s, _ := c.syncGitChartSource(repo, hr); s != nil {
+		return s, true
+	}
+	return nil, false
+}
+
+// Delete cleans up the git chart source for the given resource ID,
+// this includes the mirror if there is no reference to it from sources.
+// It returns a boolean indicating a successful removal (`true` if so,
+// `false` otherwise).
+func (c *GitChartSourceSync) Delete(hr *v1.HelmRelease) bool {
+	c.sourcesMu.Lock()
+	defer c.sourcesMu.Unlock()
+
+	// Attempt to get the source from store.
+	source, ok := c.sources[hr.ResourceID().String()]
+	if ok {
+		source.Lock()
+		defer source.Unlock()
+
+		if source.Export != nil {
+			// Clean-up the export.
+			source.Export.Clean()
+		}
+
+		// Remove the in store source.
+		delete(c.sources, hr.ResourceID().String())
+
+		if hrs, err := c.getHelmReleasesForMirror(source.Mirror); err == nil && len(hrs) == 0 {
+			// The mirror is no longer in use by any source;
+			// stop and delete the mirror.
+			c.mirrors.StopOne(source.Mirror)
+			ok = true
+		}
+	}
+	return ok
+}
+
+// SyncMirrors instructs all git mirrors to sync from their respective upstreams.
+func (c *GitChartSourceSync) SyncMirrors() {
+	c.logger.Log("info", "starting git chart source sync")
+	for _, err := range c.mirrors.RefreshAll(c.config.GitTimeout) {
+		c.logger.Log("error", "failed syncing git mirror", "err", err)
+	}
+	c.logger.Log("info", "finished syncing git chart sources")
+}
+
+// syncGitChartSource synchronizes the source in store with the latest
+// HEAD for the configured ref in the given repository. But only if it
+// has seen commits for paths we are interested in. It returns the
+// synchronized source or nil and a boolean indicating if the source
+// was updated.
+func (c *GitChartSourceSync) syncGitChartSource(repo *git.Repo, hr *v1.HelmRelease) (*GitChartSource, bool) {
+
+	mirrorName := mirrorName(hr)
+	if mirrorName == "" {
+		return nil, false
+	}
+
+	source := hr.Spec.GitChartSource
+	nsClient := c.client.HelmV1().HelmReleases(hr.Namespace)
+	logger := log.With(c.logger,
+		"mirror", source.GitURL, "ref", source.RefOrDefault(), "path", source.Path, "resource", hr.ResourceID().String())
+
+	if repoStatus, err := repo.Status(); repoStatus != git.RepoReady {
+		logger.Log("warning", "mirror not ready for sync", "status", repoStatus)
+		_ = status.SetCondition(nsClient, *hr, status.NewCondition(
+			v1.HelmReleaseChartFetched,
+			corev1.ConditionUnknown,
+			ReasonGitNotReady,
+			"mirror not ready for sync: "+err.Error(),
+		))
+		// Repository is not ready yet; wait for signal.
+		return nil, false
+	}
+
+	// Acquire sources lock and attempt to find a source in store.
+	c.sourcesMu.Lock()
+	s, ok := c.sources[hr.ResourceID().String()]
+	if !ok {
+		// No source found in store;
+		// create the boilerplate of a new one.
+		s = &GitChartSource{Mutex: sync.Mutex{}, Mirror: mirrorName, Remote: source.GitURL}
+		c.sources[hr.ResourceID().String()] = s
+	}
+	// Acquire source lock and unlock sources lock.
+	s.Lock()
```

Comment:
Locking the entry while the map lock is done means that doing the opposite can cause a deadlock. I haven't been able to see that happening in the code, but something to have in mind.